File: 0005.longest-palindromic-substring/5.longest-palindromic-substring.go
package main

/*
 * @lc app=leetcode id=5 lang=golang
 *
 * [5] Longest Palindromic Substring
 */
func longestPalindrome(s string) string {
	n := len(s)
	if n < 2 {
		return s
	}
	l, r := 0, 0
	dp := make([][]bool, n, n)
	for i := 0; i < n; i++ {
		dp[i] = make([]bool, n, n)
	}
	for j := 1; j <= n-1; j++ {
		for i := j - 1; i >= 0; i-- {
			if s[i] == s[j] && (j-i <= 2 || dp[i+1][j-1]) {
				dp[i][j] = true
				if j-i > r-l {
					l = i
					r = j
				}
			}
		}
	}
	return s[l : r+1]
}

File: 0005.longest-palindromic-substring/5.longest-palindromic-substring.py
#
# @lc app=leetcode id=5 lang=python3
#
# [5] Longest Palindromic Substring
#

# @lc code=start


class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        dp = [[False for i in range(n)] for i in range(n)]
        l, r = 0, 0
        for j in range(1, n):
            for i in range(j-1, -1, -1):
                if (s[i] == s[j]) and (j-i <= 2 or dp[i+1][j-1]):
                    dp[i][j] = True
                    if j-i > r-l:
                        l = i
                        r = j
        return s[l:r+1]
# @lc code=end
